From ffeb4270a8ab45a7e5473db94201d2211a978e66 Mon Sep 17 00:00:00 2001
From: Adam Jiang <chaoj@nvidia.com>
Date: Mon, 10 Feb 2014 19:43:33 +0900
Subject: [PATCH] Add cflags and libs for X11

---
 scripts/custom.sh | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/scripts/custom.sh b/scripts/custom.sh
index 50184c4..09eeaff 100644
--- a/scripts/custom.sh
+++ b/scripts/custom.sh
@@ -21,10 +21,11 @@
 add_yesno_var dependency "generate dependency Makefiles" no
 
 add_var xkb_list_file "XKB list file" /usr/share/X11/xkb/rules/base.lst
+add_pkg_vars x11  x11
 add_pkg_vars glib glib-2.0
 add_pkg_vars gtk  gtk+-2.0 "--atleast-version=2.12"
 
 # These vars are special - their values will be added to CFLAGS and LDFLAGS
-add_var cflagsx "C flags" '$glib_cflags $gtk_cflags -fPIC $alsa_cflags'
-add_var ldflagsx "linker flags" '$glib_libs $gtk_libs $alsa_libs'
+add_var cflagsx "C flags" '$glib_cflags $gtk_cflags -fPIC $alsa_cflags $x11_cflags'
+add_var ldflagsx "linker flags" '$glib_libs $gtk_libs $alsa_libs $x11_libs'
 
-- 
1.8.5.2

